### DISCUSSION TITLE
Cluster buildings for landcover polygons

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#771] [Q&A] [answered]
I would like to cluster buildings by distance and then use the clusters to make village and city polygons.

In QGIS, this roughly looks like this:

![image](https://github.com/onthegomap/planetiler/assets/53421382/0cbd076c-fa17-4017-8d78-5dd56cdb1991)

Here is a zoom in:

![image](https://github.com/onthegomap/planetiler/assets/53421382/5aed185d-7487-4e1b-9a07-2078ad4b2948)

The steps are:

- Import building polygons.
- Extract polygon verticies.
- Cluster the vertices by distance with DBSCAN. I used a distance threshold of 0.001 degrees.
- Make a concave hull around each cluster.
- Buffer the concave hull with a distance of 0.0001 degrees.

Is it possible to do something similar in Planetiler?

I think this could be very useful to generate village and city polygons for lower zoom levels, i.e., for a landcover dataset...

**Comment by msbarry** [comment]
Would it work to do this at the tile level? Those operations are all available through JTS, in fact they are very similar to the polygon merging code which does a dbscan clustering then buffer/union/unbuffer.  You could either pull the internal utilities out and skip the unbuffer step to get something similar, or replace the union with doing a concave hull, or to get exactly the same thing you could have the "process" step pass through building vertices, then in post process do the dbscan/concave hull/buffer steps.  You just need to make sure your tile buffer in "process" step includes enough space on either side of the tile that the shapes will line up at tile seams.

**Reply by wipfli** [reply]
I think it should be possible to do this at the tile level. At least this should work for zoom levels where the buildings are still there. But then the goal would also be to show these polygons down to zoom levels where you see for example all of Europe with one urban blob at London, Paris, and Milano.

Maybe it is easier to preprocess the buildings with something like PostGIS and generate some urban polygons first that then can be consumed by Planetiler...

**Reply by msbarry** [reply, marked as answer]
Doing it ahead of time sounds like it makes sense too. Planetiler is currently set up to efficiently do any sort of transformation on individual features (computing lake centerlines, polygon centroids, ...) any sort of transformation on many features in a rendered tile  in tile pixel coordinates (polygon merging, collapsing carriageways into a single line, ...), but it's not set up to efficiently transform many features that span many tiles.

The OpenMapTiles profile does this with the `boundary` layer and the `agg_stop` attribute on transit stop POIs by storing features during `process` and processing them in `finish` after the entire input source has been read - which works for a small number of features, but won't scale up to buildings.

It might be a good future goal for planetiler to have an efficient/scalable way to pre-process features like this before tiling them, but it would take some more thought and a new approach. Until then, if you can't do it as tile post-processing and the data is too big to fit in memory then pre-processing the data with another tool is probably your best bet.
